README.md
snapshot
----------

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Coveralls][coveralls-image]][coveralls-url]
[![David deps][david-image]][david-url]
[![node version][node-image]][node-url]
[![npm download][download-image]][download-url]
[![Gittip][gittip-image]][gittip-url]

[npm-image]: https://img.shields.io/npm/v/koa-snapshot.svg?style=flat-square
[npm-url]: https://npmjs.org/package/koa-snapshot
[travis-image]: https://img.shields.io/travis/koajs/snapshot.svg?style=flat-square
[travis-url]: https://travis-ci.org/koajs/snapshot
[coveralls-image]: https://img.shields.io/coveralls/koajs/snapshot.svg?style=flat-square
[coveralls-url]: https://coveralls.io/r/koajs/snapshot?branch=master
[david-image]: https://img.shields.io/david/koajs/snapshot.svg?style=flat-square
[david-url]: https://david-dm.org/koajs/snapshot
[node-image]: https://img.shields.io/badge/node.js-%3E=_0.11-red.svg?style=flat-square
[node-url]: http://nodejs.org/download/
[download-image]: https://img.shields.io/npm/dm/koa-snapshot.svg?style=flat-square
[download-url]: https://npmjs.org/package/koa-snapshot
[gittip-image]: https://img.shields.io/gittip/dead-horse.svg?style=flat-square
[gittip-url]: https://www.gittip.com/dead-horse/

take snapshot when request, cache by request path.

## Install

```
npm install koa-snapshot
```

## Usage

```
var koa = require('koa');
var snapshot = require('koa-snapshot');
var app = koa();

app.use(snapshot());

```

## Options

Use your own cache client by `options.cache`.
If do not present `options.cache`, snapshot will create a [lru-cache](https://github.com/isaacs/node-lru-cache) instance with the options.

default options for lru-cache:

```
{
  max: 50 * 1024 * 1024,  // 50mb
  maxAge: 12 * 60 * 60 * 1000,  // 12h
  length: function (n) {
    return n.length;
  }
}
```

## Turn off Snapshot

You can manually turn off snapshot by set `this.noSnapshot = true` in every request.

## License

MIT


History.md

1.2.0 / 2014-11-06
==================

  * support non-throw 5xx get from cache

1.1.1 / 2014-09-26
==================

  * add coverage
  * fix response type cache

1.1.0 / 2014-08-19
==================

  * support manually turn off snapshot

1.0.3 / 2014-07-11
==================

  * support options.cache
  * Release 1.0.3

1.0.2 / 2014-06-28
==================

  * fix length

1.0.1 / 2014-06-28
==================

  * cache content type for charset

1.0.0 / 2014-06-27
==================

  * set header X-Snapshot-Status

0.1.0 / 2014-06-25
==================

  * init


test/snapshot.test.js
/*!
 * snapshot - test/snapshot.test.js
 * Copyright(c) 2014 dead_horse <dead_horse@qq.com>
 * MIT Licensed
 */

'use strict';

/**
 * Module dependencies.
 */

var request = require('supertest');
var should = require('should');
var snapshot = require('..');
var koa = require('koa');

var app = koa();
app.use(snapshot());

var status = 200;
var noSnapshot = false;
app.use(function* () {
  this.noSnapshot = noSnapshot;
  if (status === 200) {
    this.type = 'text/html; charset=gbk';
    return this.body = 'hello world';
  }
  if (status === 201) {
    this.status = 201;
    return this.body = {ok: true};
  }
  if (status === 302) {
    return this.redirect('https://github.com');
  }
  // 501 don't throw
  if (status === 501) {
    return this.status = 501;
  }
  this.throw(status);
});

app = app.callback();

describe('test/snapshot.test.js', function () {
  afterEach(function () {
    status = 200;
    noSnapshot = false;
  });

  it('should 500', function (done) {
    status = 500;
    request(app)
    .get('/')
    .expect(500, done);
  });

  describe('cache html', function () {
    before(function (done) {
      request(app)
      .get('/')
      .expect(200)
      .expect('hello world', done);
    });

    it('should return cached html when throw error', function (done) {
      noSnapshot = false;
      status = 500;
      request(app)
      .get('/')
      .expect(200)
      .expect('X-Snapshot-Status', 500)
      .expect('Content-Type', 'text/html; charset=gbk')
      .expect('hello world', done);
    });

    it('should return cached html when status > 500', function (done) {
      noSnapshot = false;
      status = 501;
      request(app)
      .get('/')
      .expect(200)
      .expect('X-Snapshot-Status', 501)
      .expect('Content-Type', 'text/html; charset=gbk')
      .expect('hello world', done);
    });

    it('should not cache when 302', function (done) {
      status = 302;
      request(app)
      .get('/')
      .expect(302, done);
    });

    it('should not cache when 404', function (done) {
      status = 404;
      request(app)
      .get('/')
      .expect(404, done);
    });

    it('should not cache when not html', function (done) {
      status = 201;
      request(app)
      .get('/')
      .expect({ok: true})
      .expect(201, done);
    });

    it('should not use cache when < 500', function (done) {
      status = 404;
      request(app)
      .get('/')
      .expect(404, done);
    });

    it('should not use cache when noSnapshot = true', function (done) {
      status = 500;
      noSnapshot = true;
      request(app)
      .get('/')
      .expect(500, done);
    });

    it('should not cache when noSnapshot = true', function (done) {
      noSnapshot = true;
      request(app)
      .get('/nocache')
      .expect(200, function (err) {
        should.not.exist(err);
        status = 500;
        noSnapshot = false;
        request(app)
        .get('/nocache')
        .expect(500, done);
      });
    });
  });
});


index.js
/*!
 * snapshot - index.js
 * Copyright(c) 2014 dead_horse <dead_horse@qq.com>
 * MIT Licensed
 */

'use strict';

/**
 * Module dependencies.
 */

var copy = require('copy-to');

var defaultOptions = {
  max: 50 * 1024 * 1024,  // 50mb
  maxAge: 12 * 60 * 60 * 1000,  // 12h
  length: function (n) {
    return n.body ? n.body.length : 0;
  }
};

module.exports = function (options) {
  options = options || {};
  copy(defaultOptions).to(options);

  var cache = options.cache || require('lru-cache')(options);

  return function* snapshot(next) {
    try {
      yield *next;
    } catch (err) {
      return errorHandler(err, this);
    }
    getFromCache(this);
    setCache(this);
  };

  function setCache(ctx) {
    if (ctx.noSnapshot) return;
    // only cache html
    if (!ctx.response.is('html')) return;
    // only cache 2xx
    if ((ctx.status / 100 | 0) !== 2) return;
    // update cache
    cache.set(ctx.path, {
      body: ctx.body,
      type: ctx.response.get('Content-Type') || 'html'
    });
  }

  function getFromCache(ctx) {
    if (ctx.noSnapshot) return;
    if (ctx.status < 500) return;
    var c = cache.get(ctx.path);
    if (!c) return;

    ctx.type = c.type;
    ctx.body = c.body;
    ctx.set('X-Snapshot-Status', ctx.status || 500);
    ctx.status = 200;
    return;
  }

  function errorHandler(err, ctx) {
    if (ctx.noSnapshot) throw err;
    if (err.status && err.status < 500) throw err;
    var c = cache.get(ctx.path);
    if (!c) throw err;

    ctx.app.emit('error', err);
    ctx.type = c.type;
    ctx.body = c.body;
    ctx.set('X-Snapshot-Status', err.status || 500);
    return;
  }

};
